src/language-markdown/print-whitespace.js

```diff
@@ -0,0 +1,310 @@
+      if (
+        (previousKind === KIND_CJ_LETTER && nextKind === KIND_NON_CJK) ||
+        (previousKind === KIND_NON_CJK && nextKind === KIND_CJ_LETTER)
+      ) {
+        ++stats[node.value];
+      }
+    }
+  }
+
+  // Injects a property to cache the result.
+  sentenceNode.usesCJSpaces = stats[" "] > stats[""];
+  return sentenceNode.usesCJSpaces;
+}
+
+/**
+ * @typedef {import("./utils.js").WordNode} WordNode
+ * @typedef {import("./utils.js").WhitespaceValue} WhitespaceValue
+ * @typedef {{next?: WordNode | undefined | null, previous?: WordNode | undefined | null}} AdjacentNodes
+ * Adjacent node to `WhitespaceNode`. the consecution of `WhitespaceNode` is a bug, so adjacent nodes must be `WordNode`.
+ * @typedef {import("./utils.js").WordKind} WordKind
+ * @typedef {import("../common/ast-path.js").default} AstPath
+ */
+
+/**
+ * Checks if given `"\n"` node can be converted to Space
+ *
+ * For example, if you would like to squash the multi-line string `"You might want\nto use Prettier."` into a single line,
+ * you would replace `"\n"` with `" "`. (`"You might want to use Prettier."`)
+ *
+ * However, you should note that Chinese and Japanese do not use U+0020 Space to divide words, so U+000A End of Line must not be replaced with it.
+ * Behavior in other languages (e.g. Thai) will not be changed because there are too much things to consider. (PR welcome)
+ *
+ * @param {AstPath} path path of given node
+ * @param {AdjacentNodes | undefined} adjacentNodes adjacent sibling nodes of given node
+ * @returns {boolean} `true` if given node can be convertedToSpace, `false` if not (i.e. newline or empty character)
+ */
+function canBeConvertedToSpace(path, adjacentNodes) {
+  // no adjacent nodes
+  if (!adjacentNodes) {
+    return true;
+  }
+
+  // e.g. " \nletter"
+  if (!adjacentNodes.previous || !adjacentNodes.next) {
+    return true;
+  }
+
+  const previousKind = adjacentNodes.previous.kind;
+  const nextKind = adjacentNodes.next.kind;
+
+  // "\n" between not western or Korean (han, kana, CJK punctuations) characters always can be converted to Space
+  // Korean hangul simulates latin words; see #6516 (https://github.com/prettier/prettier/issues/6516)
+  if (
+    isWesternOrKoreanLetter(previousKind) &&
+    isWesternOrKoreanLetter(nextKind)
+  ) {
+    return true;
+  }
+
+  // han & hangul: same way preferred
+  if (
+    (previousKind === KIND_K_LETTER && nextKind === KIND_CJ_LETTER) ||
+    (nextKind === KIND_K_LETTER && previousKind === KIND_CJ_LETTER)
+  ) {
+    return true;
+  }
+
+  // Do not convert it to Space when:
+  if (
+    // Shall not be converted to Space around CJK punctuation
+    previousKind === KIND_CJK_PUNCTUATION ||
+    nextKind === KIND_CJK_PUNCTUATION ||
+    // "\n" between CJ always SHALL NOT be convertedToSpace
+    (isCJK(previousKind) && isCJK(nextKind))
+  ) {
+    return false;
+  }
+
+  const characterBefore = adjacentNodes.previous.value.at(-1);
```

Comment:
Will this index access cause problems for emojis or other characters
? https://mathiasbynens.be/notes/javascript-unicode

Reply:
Good question. I checked and all the characters in `lineBreakBetweenTheseAndCJKConvertsToSpace` have `.length === 1`. As long as it's the case that index access isn't a problem.

Reply:
@fisker As all punctuation marks we at are concerned with are in BMP, no problems about this will not happen at least for the time being.